scio-avro/src/main/scala/com/spotify/scio/avro/AvroDatumFactory.scala

```diff
@@ -16,27 +16,97 @@
 
 package com.spotify.scio.avro
 
-import org.apache.avro.Schema
+import org.apache.avro.{Conversion, Schema}
+import org.apache.avro.generic.{GenericDatumReader, GenericRecord}
 import org.apache.avro.io.{DatumReader, DatumWriter}
-import org.apache.avro.reflect.{ReflectData, ReflectDatumReader, ReflectDatumWriter}
+import org.apache.avro.specific.{SpecificDatumReader, SpecificDatumWriter, SpecificRecord}
 import org.apache.beam.sdk.extensions.avro.io.AvroDatumFactory
-import org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils
 
 /**
- * Custom AvroDatumFactory for avro AvroDatumFactory relying on avro reflect so that underlying
- * CharSequence type is String
+ * AvroDatumFactory for [[GenericRecord]] forcing underlying [[CharSequence]] implementation to
+ * [[String]]
+ *
+ * Avro default [[CharSequence]] implementation is [[org.apache.avro.util.Utf8]] which can't be used
+ * when joining or as SMB keys as it doest not implement equals
  */
-private[scio] class SpecificRecordDatumFactory[T](recordType: Class[T])
-    extends AvroDatumFactory[T](recordType) {
-  override def apply(writer: Schema, reader: Schema): DatumReader[T] = {
-    val data = new ReflectData(recordType.getClassLoader)
-    AvroUtils.addLogicalTypeConversions(data)
-    new ReflectDatumReader[T](writer, reader, data)
+private[scio] object GenericRecordDatumFactory extends AvroDatumFactory.GenericDatumFactory {
+
+  private class ScioGenericDatumReader extends GenericDatumReader[GenericRecord] {
+    override def findStringClass(schema: Schema): Class[_] = super.findStringClass(schema) match {
+      case cls if cls == classOf[CharSequence] => classOf[String]
+      case cls                                 => cls
+    }
+  }
+  override def apply(writer: Schema, reader: Schema): DatumReader[GenericRecord] = {
+    val datumReader = new ScioGenericDatumReader()
+    datumReader.setExpected(reader)
+    datumReader.setSchema(writer)
+    datumReader
+  }
+}
+
+/**
+ * AvroDatumFactory for [[SpecificRecord]] forcing underlying [[CharSequence]] implementation to
+ * [[String]]
+ *
+ * Avro default [[CharSequence]] implementation is [[org.apache.avro.util.Utf8]] which can't be used
+ * when joining or as SMB keys as it doest not implement equals
+ */
+private[scio] class SpecificRecordDatumFactory[T <: SpecificRecord](recordType: Class[T])
+    extends AvroDatumFactory.SpecificDatumFactory[T](recordType) {
+  import SpecificRecordDatumFactory._
```

Comment:
Maybe I'm just reading the diff wrong, but are these imports actually needed in the `SpecificRecordDatumFactory` class?

Reply:
Just getting access to `runtimeAvroVersion` and `specificRecordConversions` methods from the companion object.